src/GovUk.Education.ExploreEducationStatistics.Admin/Services/DataBlockMigrationService.cs

```diff
@@ -1,40 +1,456 @@
+
+                                    var createdDate = await GetDataBlockCreatedDate(dataBlock);
+
+                                    var boundaryLevel = GetLatestBoundaryLevelAtCreationTime(
+                                        createdDate,
+                                        nonCountryGeographicLevel,
+                                        boundaryLevels);
+
+                                    dataBlock.Query.BoundaryLevel = boundaryLevel.Id;
+                                    map.BoundaryLevel = boundaryLevel.Id;
+                                    await SaveChanges(dataBlock, dryRun);
+                                    return new MapMigrationResult(
+                                        plan.DataBlockId,
+                                        plan.Type,
+                                        plan.Strategy,
+                                        map.BoundaryLevel,
+                                        boundaryLevels.Single(level => level.Id == map.BoundaryLevel).Level);
+                                }
+                                case QuestionWhichCorrectGeoLevelToSelect:
+                                    return new MapMigrationResult(
+                                        plan.DataBlockId,
+                                        plan.Type,
+                                        plan.Strategy,
+                                        null,
+                                        null);
+                                case FixMapConfigNoDataSetGeographicLevels:
+                                    return new MapMigrationResult(
+                                        plan.DataBlockId,
+                                        plan.Type,
+                                        plan.Strategy,
+                                        null,
+                                        null);
+                                case FixMapConfigInvalidDataSetGeographicLevels:
+                                    return new MapMigrationResult(
+                                        plan.DataBlockId,
+                                        plan.Type,
+                                        plan.Strategy,
+                                        null,
+                                        null);
+                                default:
+                                    throw new ArgumentOutOfRangeException(
+                                        $"Unknown migration strategy {plan.Strategy}");
+                            }
+                        })
+                        .ToListAsync();
+
+                    return migrationResults;
                 });
         }
+
+        private async Task SaveChanges(DataBlock dataBlock, bool dryRun)
+        {
+            if (!dryRun)
+            {
+                _contentDbContext.DataBlocks.Update(dataBlock);
+                await _contentDbContext.SaveChangesAsync();
```

Comment:
Would it be possible to call `SaveChangesAsync` once at the end to prevent the migration from partially running?

Reply:
I'm happy to allow a partial migration as it's rerunnable.  It'll only attempt to migrate Maps that aren't yet migrated, so happy if a partial failure doesn't roll back all the successful operations to date.